### PR TITLE
Add rust support

### PR DESCRIPTION
**Pull request by AleksaC**
I mostly followed the steps from the [official buster image](https://github.com/rust-lang/docker-rust/blob/master/1.48.0/buster/Dockerfile). Besides the minimal profile I added `rustfmt` and `clippy` as I'm not sure they can be installed properly through additional dependencies. Adding rust toolchain increases the image size by `~45%`.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -87,4 +87,23 @@ RUN : \
...
+RUN : \
+    && rustArch='x86_64-unknown-linux-gnu' \
+    && curl --silent --location --output rustup-init "https://static.rust-lang.org/rustup/archive/${RUSTUP_VERSION}/${rustArch}/rustup-init" \
+    && echo "${RUSTUP_SHA256} rustup-init" | sha256sum --check \
+    && chmod +x rustup-init \
+    && ./rustup-init -y --profile minimal --no-modify-path --default-toolchain $RUST --default-host ${rustArch} \
+    && chmod -R a+w $RUSTUP_HOME $CARGO_HOME \
+    && rm -rf rustup-init \
+    && rustup component add rustfmt \
+    && rustup component add clippy \
```
Comment (asottile) [inline Dockerfile:106]: these can be installed through `cargo` so they shouldn't be here

File: Dockerfile [hunk 2]
```diff
@@ -87,4 +87,23 @@ RUN : \
...
+    RUSTUP_HOME=/usr/local/rustup \
+    CARGO_HOME=/usr/local/cargo \
+    PATH=/usr/local/cargo/bin:$PATH
+RUN : \
+    && rustArch='x86_64-unknown-linux-gnu' \
+    && curl --silent --location --output rustup-init "https://static.rust-lang.org/rustup/archive/${RUSTUP_VERSION}/${rustArch}/rustup-init" \
+    && echo "${RUSTUP_SHA256} rustup-init" | sha256sum --check \
+    && chmod +x rustup-init \
+    && ./rustup-init -y --profile minimal --no-modify-path --default-toolchain $RUST --default-host ${rustArch} \
+    && chmod -R a+w $RUSTUP_HOME $CARGO_HOME \
```
Comment (asottile) [inline Dockerfile:103]: these should be quoted (both this line and the one above it) even though the variables currently don't matter

also `${rustArch}` (1) should be quoted and (2) the braces aren't doing anything

**Review by asottile — COMMENTED** [review]
please add this to the info script as well so that the rust / cargo versions are reported there

ideally we'd also add a test to demonstrate that rust hooks work (see `bin/test`)

File: Dockerfile (hunk 2 above)
Comment (asottile) [inline Dockerfile:103]: also these should not be global writable, so actually just remove this line

**Comment by AleksaC** [comment]
I fixed the problems above and will also add a test.

File: bin/_info [hunk 3]
```diff
@@ -89,6 +89,16 @@ def main() -> int:
     print()
     with _console():
         _call('ruby', '--version')
+    print()
+
+    print('## rust')
+    print()
+    with _console():
+        _call('rustup', '--version')
+        print()
```
Comment (asottile) [inline bin/_info:98]: I don't think they'll be able to run `rustup` so just cargo and rustc are relevant here? (though admittedly I'm not super knowledgeable in this space so I could be wrong here)

Comment (AleksaC) [reply bin/_info:98]: I'm not sure what do you mean here. One possible use of rustup that comes to mind is if somebody used this as a base image and wants to install another toolchain.

Comment (asottile) [reply bin/_info:98]: that's not a use case I intend to support so it's not worth documenting

**Comment by AleksaC** [comment]
I need to make at least the `$CARGO_HOME/registry` writable because it is needed for crates.io index. 

Also it seems like `clippy` is no longer available through crates.io. The situation with `rustfmt` is similar, there is a crate [`rustfmt-nightly`](https://crates.io/crates/rustfmt-nightly) but `--force` flag needs to be passed to cargo, and it also failed to build for both stable and nightly toolchain when I tried it. So it seems that the only way to build these two through cargo is from source, and that would require nightly toolchain. 

I cannot find any pre-commit hooks using rust other than [these](https://github.com/doublify/pre-commit-rust) which use system installations, so I created a trivial one for the test. It's a bit hacky because the repo uses itself in `.pre-commit-config.yml` but IMO it is still better than creating two trivial repos. Another option would be to create a remote trivial rust hook and use it instead.

**Comment by asottile** [comment]
> 
> 
> I need to make at least the `$CARGO_HOME/registry` writable because it is needed for crates.io index.
> 
> Also it seems like `clippy` is no longer available through crates.io. The situation with `rustfmt` is similar, there is a crate [`rustfmt-nightly`](https://crates.io/crates/rustfmt-nightly) but `--force` flag needs to be passed to cargo to install it, and it also failed to build for both stable and nightly toolchain when I tried it. So it seems that the only way to build these two through cargo is from source, and that would require nightly toolchain.
> 
> I cannot find any pre-commit hooks using rust other than [these](https://github.com/doublify/pre-commit-rust) which use system installations, so I created a trivial one for the test. It's a bit hacky because the repo uses itself in `.pre-commit-config.yml` but IMO it is still better than creating two trivial repos. Another option would be to create a remote trivial rust hook and use it instead.

hmmm that is quite unfortunate re: clippy / rustfmt -- they both worked from crates when rust support was added to pre-commit -- I guess we can add those components here but it feels wrong

note that pre-commit's rust support does include `local` + `additional_dependencies` -- perhaps you're not using the special syntax? https://pre-commit.com/#rust

**Comment by AleksaC** [comment]
Do you mean that I could just drop a random crate in additional dependencies for the test?

**Comment by asottile** [comment]
> 
> 
> Do you mean that I could just drop a random crate in additional dependencies for the test?

yes

**Comment by asottile** [comment]
pre-commit itself uses `cli:shellharden:3.1.0`

**Comment by AleksaC** [comment]
Haven't thought of that for some reason 😅, I'll do that.

**Review by asottile — APPROVED** [review]

**Pull request merged**